File: archive/ruby/muse/data_structures/LinkedStack.rb
module LinkedStack
  class Node
    attr_accessor :data, :next

    def initialize(element)
      @data = element
      @next = nil
    end
  end

  class LinkedStack
    def initialize
      @head = nil
      @tail = nil
      @length = 0
    end

    def size
      @length
    end

    def isEmpty
      @length.zero?
    end

    def peek
      raise "[PANIC - NoSuchElement]" if @length.zero?

      @tail.data
    end

    def push(element)
      node = Node.new(element)

      if @length.zero?
        @head = node
      else
        @tail.next = node
      end

      @tail = node

      @length += 1
    end

    def pop
      raise "[PANIC - NoSuchElement]" if @length.zero?

      target = @tail

      if @length == 1
        @head = nil
        @tail = nil
      else
        cursor = @head
        (@length - 2).times do
          cursor = cursor.next
        end
        cursor.next = nil
        @tail = cursor
      end

      target.data = nil

      @length -= 1
    end
  end
end

File: archive/nodejs/muse/data_structures/LinkedStack.js
class Node {
  constructor(element) {
    this.data = element;
    this.next = null;
  }
}

export class LinkedStack {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  size() {
    return this.length;
  }

  isEmpty() {
    return this.length === 0;
  }

  peek() {
    if (this.length === 0) {
      throw new Error('[PANIC - NoSuchElement]');
    }

    return this.tail.data;
  }

  push(element) {
    const node = new Node(element);

    if (this.length === 0) {
      this.head = node;
    } else {
      this.tail.next = node;
    }

    this.tail = node;

    this.length++;
  }

  pop() {
    if (this.length === 0) {
      throw new Error('[PANIC - NoSuchElement]');
    }

    const target = this.tail;

    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      let cursor = this.head;
      for (let i = 0, bound = this.length - 2; i < bound; i++) {
        cursor = cursor.next;
      }
      cursor.next = null;
      this.tail = cursor;
    }

    target.data = null;

    this.length--;
  }
}
